File: tests/python/scripts/rules/T019.py
#!/usr/bin/python
# control structures should have complete curly-braced block of code

tokenTypes = ["for", "if", "while", "do", "else", "leftparen", "rightparen", "leftbrace", "rightbrace", "semicolon"]
START, CONTROL, EXPECTED_BLOCK, BLOCK = range(4)

for fileName in vera.getSourceFileNames():
  state = START
  prev = None
  for t in vera.getTokens(fileName, 1, 0, -1, -1, tokenTypes):
    if state == CONTROL:
      if t.type == "leftparen":
        parenCount += 1
      elif t.type == "rightparen":
        parenCount -= 1
        if parenCount == 0:
          state = EXPECTED_BLOCK
    elif state == EXPECTED_BLOCK:
      if prev == "else" and t.type == "if":
        pass
      elif t.type != "leftbrace":
        vera.report(fileName, t.line, "full block {} expected in the control structure")
      state = BLOCK

    if t.type == "for" or t.type == "if":
      parenCount = 0
      state = CONTROL
    elif t.type == "do" or t.type == "else":
      state = EXPECTED_BLOCK
    elif t.type == "while" and prev != "rightbrace":
      parenCount = 0
      state = CONTROL
    prev = t.type

File: tests/lua/rules/T019.lua
-- control structures should have complete curly-braced block of code

filter = {"for", "if", "while", "do", "else", "leftparen", "rightparen", "leftbrace", "rightbrace", "semicolon"}
START, CONTROL, EXPECTED_BLOCK, BLOCK = 0, 1, 2, 3

for fileName in getSourceFileNames() do
  state = START
  prev = nil
  for t in getTokens(fileName, 1, 0, -1, -1, filter) do
    if state == CONTROL then
      if t.type == "leftparen" then
        parenCount = parenCount + 1
      elseif t.type == "rightparen" then
        parenCount = parenCount - 1
        if parenCount == 0 then
          state = EXPECTED_BLOCK
        end
      end
    elseif state == EXPECTED_BLOCK then
      if prev == "else" and t.type == "if" then
        -- do nothing
      elseif t.type ~= "leftbrace" then
        report(fileName, t.line, "full block {} expected in the control structure")
      end
      state = BLOCK
    end

    if t.type == "for" or t.type == "if" then
      parenCount = 0
      state = CONTROL
    elseif t.type == "do" or t.type == "else" then
      state = EXPECTED_BLOCK
    elseif t.type == "while" and prev ~= "rightbrace" then
      parenCount = 0
      state = CONTROL
    end
    prev = t.type
  end
end
